### ISSUE TITLE
Prolate spheroid model with self-demag

### ISSUE THREAD

**Issue by drandykass** [fatiando/fatiando#258]
Is there any interest in having a prolate spheroid analytic mag solution which incorporates self-demagnetization?  I need to write one up to compare some results, and could do it 'fatiando-ish' if desired.

**Comment by drandykass** [comment]
You know, that could really go well with the 'seeded' inversion that you and Valeria published a while ago.  I forget what you called it.

**Comment by leouieda** [comment]
@drandykass yes please! Any new forward modeling is always welcome. 

How so? Do you mean using the spheroids to build  the model?

**Comment by drandykass** [comment]
Indeed.  Instead of using spheres, you use spheroids.  You could then build in anisotropy or self-demag into the solution.  Forgive me--it's been ages since I read that paper so I could be way off.

**Comment by leouieda** [comment]
No problem! In that case it wouldn't work probably because there would be a lot of empty spaces when trying to aggregate spheroids.

**Comment by drandykass** [comment]
Ah, I didn't realize that it was trying to fill the space.  I'll have to go back and reread it.

**Comment by drandykass** [comment]
I was looking through to see if there was already a ticket for an issue I was having and found this very old comment.  Guess I need to get on that...

I have the code written, I just need to make it Fatiando-ish!

**Comment by leouieda** [comment]
No hurry. I've been neglecting Fatiando for past few months.

I've been experimenting with stripping away some of the parts that aren't necessarily "geophysics", like the inversion package and `mesher`. These would become separate libraries and Fatiando would only have the geophysics code.

I started with `mesher` and made [fatiando/geometric](https://github.com/fatiando/geometric/blob/master/geometric/base.py). I've been rethinking the way that the physical properties should be represented. The `props` dict is a bad idea and doesn't support validation of properties. For example, if given density, the class should know how to calculate volume and mass. Or given a mass, it should automatically fill out the density. From conversations with the Simpeg team, it seems that [traitlets](https://github.com/ipython/traitlets) are a good fit.

This would also help us share code with their [discretize package](https://github.com/simpeg/discretize).

**Comment by drandykass** [comment]
Indeed.  Somewhat related, but with important implications, is something I
wanted to add.  I needed a forward model of a cylinder with uniform
magnetization (infinitely extending and horizontal, but with magnetization
not parallel to the axis).  Right now, that should be a new object in
mesher (e.g. fatiando.mesher.infCylinder or something).  Should I not
proceed with that?  I can just write my own little script.  If you would
like it added, what would you suggest I call it?  HorizInfCylinder doesn't
have a nice ring to it...

**Comment by leouieda** [comment]
You can add it to mesher for now and we'll move it over to the new package once that is on its way.

So this is a horizontal cylinder that is infinitely long? Can you control the orientation of the cylinder? Is it truly 3D (e.g., can the magnetization be in any direction)?

**Comment by drandykass** [comment]
Understood.

Yes, this is horizontal cylinder of arbitrary radius, angled with whatever declination.  The magnetisation can be in any direction (of course B is zero if aligned with axis), but must be constant throughout.  So the physical parameters in geometry are declination, depth to centre of cylinder, radius, and magnetisation, both magnitude and direction.  It makes more sense physically to cast the magnetisation as a dipole moment per unit length, but we can leave it as a magnetisation per unit volume.

The salient math comes from Blakely.  I had struggled with solving the surface integral from the vector potential formulation (such as from Griffiths), but Blakely rather simply and elegantly uses Poisson's relation to convert from the expression of the gravity potential of a horizontal cylinder.

I also dug up my prolate spheroid code, so I can add that in too (separate branch of course).
